src/darwin/Framework/CHIP/MTRDevice.h

```diff
@@ -337,6 +338,27 @@ MTR_AVAILABLE(ios(16.1), macos(13.0), watchos(9.1), tvos(16.1))
                     queue:(dispatch_queue_t)queue
                completion:(void (^)(NSURL * _Nullable url, NSError * _Nullable error))completion
     MTR_AVAILABLE(ios(17.6), macos(14.6), watchos(10.6), tvos(17.6));
+
+/**
+ * Wait until a set of attributes reaches certain values or the provided timeout
+ * (if any) is reached.  The completion is called at that point, with an
+ * MTREErrorCodeTimeout error if a timeout occurred.
+ *
+ * If the MTRAttributeValueWaiter is canceled or is destroyed before the
+ * completion is called, the completion will be called with an
+ * MTRErrorCodeCancelled error.
+ *
+ * The attributes and values to wait for are represented as a dictionary which
+ * has the attribute paths as keys and the expected data-values as values.
+ *
+ * Nil will be returned if for some reason the wait could not be started
+ * (e.g. invalid input).
+ */
+- (MTRAttributeValueWaiter * _Nullable)waitForAttributeValues:(NSDictionary<MTRAttributePath *, NSDictionary<NSString *, id> *> *)values
```

Comment:
If this is nullable, then it needs to return an NSError ** out param. 

Per our discussion as well, I feel this is mixing two worlds together... I prefer delegate, but not a hill i'll die on :)

Reply:
So I talked about the delegate idea to some of the planned API consumer, and it has a drawback, which is that it makes it hard to take different actions depending on which set of attributes you were waiting for; you sort of have to define a new protocol implementation for each new thing you decide you might need, and if you need to change the behavior at runtime at all you end up having to do blocks for the behavior anyway, but now held by the delegate.  So just in terms of how people want to use this API it does not feel like a great fit.

Good point about NSError**; if this stays nullable (as opposed to switching to "just assert on bad input"), I will do that.

Reply:
Personally, completions (pun intended) are a trap here, but if that's what the people want, sounds fine. :)

Reply:
Done.